### DISCUSSION TITLE
Guide for custom board bringup

### DISCUSSION THREAD

**Discussion by apbr** [OE4T/meta-tegra#1455] [Q&A] [answered]
Hi,
is there a guide on how to create a new MACHNE for a custom carrier board (Orin NX in this case)?
I'm looking for something that explains how to apply the instructions from the [Jetson Module Adaptation and Bring-Up](https://docs.nvidia.com/jetson/archives/r36.2/DeveloperGuide/HR/JetsonModuleAdaptationAndBringUp.html) Nvidia guide to the Yocto/meta-tegra world.
I found [Creating a custom MACHINE - meta-tegra wiki](https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE) but it seems lacking a lot of steps.

If such guide does not exist yet, I'd be glad to extend [Creating a custom MACHINE - meta-tegra wiki](https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE) if this is desired.  
(As soon as I figured out which steps are required ofc)

I am confident I figured out most steps, which seems to be the following:
- Create a new machine config (Or copy & modify an existing).
- Create a new `flashvars` file for this machine. (Or copy & modify an existing).
- Add and include the dtsi files that are produced by the [Nvidia pinmux Excel sheet](https://developer.nvidia.com/downloads/jetson-orin-nx-and-orin-nano-series-pinmux-config-template).
- Add and use a custom device tree.
- (Optionally) Disable EEPROM usage by modifiying MB2 config which is set in `MB2BCT_CFG` in `flashvars`

Do you see anything missing?

What I'm not sure about at the moment is the following:
What should I do with the dtsi files from the pinmux?
Do I just have to append `tegra-bootfiles_35.4.1.bb` so those files are placed in `${D}${datadir}/tegraflash/` and then change the `flashvars` file?
  - I guess the `pinmux.dtsi` is what is set in `PINMUX_CONFIG`.
  - `gpio.dtsi` seems to be included in `pinmux.dtsi`, so I guess I don't need to specify it anywhere else?
  - It seems `padvoltage.dtsi` is what is set in `PMC_CONFIG`.

Is this correct?
Is there any other place where I have to place or reference these new dtsis?

**Comment by madisongh** [comment, marked as answer]
Sounds good overall.  It would be great to have more detail in the wiki about this process.

> Do I just have to append tegra-bootfiles_35.4.1.bb so those files are placed in ${D}${datadir}/tegraflash/ and then change the flashvars file?

Yeah, something like that. What I've done for my custom machines is to create a new recipe for the config files generated out of the pinmux spreadsheet and have a bbappend for `tegra-bootfiles` depend on that.  The other recipe just fetched the files from a git repo where I stored them after running through the macros in the spreadsheet.

For the flashvars file, I add a bbappend for the `tegra-flashvars` recipe, which prepends to FILESEXTRAPATHS to get the custom copy of the flashvars file for my new machine.

I haven't gone through this process myself for an Orin-based machine, so I'm not 100% sure about the `.dtsi` files you're asking about, but from what I can see that looks right.

**Reply by apbr** [reply]
Ok, thanks a lot. I will do a writeup for the wiki within the next 2 weeks.

**Reply by apbr** [reply]
FYI, as promised I extended the wiki: https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE#jetson-orin
Feel free to fix any errors.

**Reply by spb06** [reply]
Hey there, thanks for much for extended this WIKI. I am going through something very similar involving a custom carrier board (AGX Orin). 

We followed the instructions in the "[Create a Custom Machine](https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE#custom-device-tree)". Specifically, we created a new machine config and flash config as advised in the wiki.

For the machine conf file, we copied over the [jetson-agx-orin-devkit.conf](https://github.com/OE4T/meta-tegra/blob/master/conf/machine/jetson-agx-orin-devkit.conf) and renamed it to match our machine name. The carrier board vendor provided their own .dtb file that we believe should replace the .dtb file specified in [jetson-agx-orin-devkit.conf](https://github.com/OE4T/meta-tegra/blob/master/conf/machine/jetson-agx-orin-devkit.conf) but we are not sure the best way to go about doing that.

Any guidance will be greatly appreciated!

**Reply by madisongh** [reply]
Device trees are normally built with the kernel, but we do have a mechanism for providing them from an alternative location. There's an example of using this [here](https://github.com/OE4T/tegra-demo-distro/commit/2215ed0c77139595d22c504df1e8ddb01c4b2b1d). That recipe builds the device trees from sources, but if you have only DTB files, you could just have the recipe directly install those.

**Reply by spb06** [reply]
Thanks for the response! 

I came across the link that you shared and tried it out but still no luck. Yocto builds and I can see my-machine.dtb, padvoltage dtsi, gpio dtsi, and pinmux dtsi in the tar file that gets generated, however, when I flash the board does not boot up completely. My guess is I am close and need to tweak a couple of things to make it work. I am running kirkstone btw. 

Essentially, below are the steps I took to add the device tree:
 1) added the following to the local.conf: 
`PREFERRED_PROVIDER_virtual/dtb = "my_machine-devicetree"`
`DTBFILE:my_machine = "my_machine.dtb"`

2) added the `my_machine.dts` (decompiled version of the .dtb that the carrier board vendor provided) under recipes-bsp/my_machine-devicetree/files

3) Coped the .bb from [here](https://github.com/OE4T/tegra-demo-distro/commit/2215ed0c77139595d22c504df1e8ddb01c4b2b1d). and changed out the name of the .dts file. 

4) Changed the KERNEL_DEVICETREE in my_machine.conf from `KERNEL_DEVICETREE ?= "tegra234-p3701-0000-p3737-0000.dtb tegra234-p3701-0004-p3737-0000.dtb"` (which was copied over from jetson-agx-orin-devkit.conf) to `KERNEL_DEVICETREE ?= ""`

My questions: 
1) Do I need to edit anything else in the .bb from [here](https://github.com/OE4T/tegra-demo-distro/commit/2215ed0c77139595d22c504df1e8ddb01c4b2b1d)? 

2) I used the decompiled .dts version of the .dtb that vendor sent me. Is there any way to use the the .dtb instead of the .dts? I tried to change the KERNEL_DEVICETREE in my_machine.conf from `KERNEL_DEVICETREE ?= "tegra234-p3701-0000-p3737-0000.dtb tegra234-p3701-0004-p3737-0000.dtb"` (which was copied over from jetson-agx-orin-devkit.conf) to `KERNEL_DEVICETREE ?= "my_machine.dtb"` and placed the my_machine.dtb file in recipes-bsp/my_machine-devicetree/files but the build failed and the error message stated "could not find recipe for target my_machine.dtb".

**Reply by madisongh** [reply]
As I mentioned above, you should be able to list the `.dtb` file(s) in the SRC_URI and just install them directly, instead of decompiling/recompiling.

As far as debugging the boot problem goes, monitoring the serial console output during the flashing and boot process, to see where it gets stuck, would be the first step.

**Reply by spb06** [reply]
Thanks again! I went ahead and tried the following and got the following error: 

- Made `KERNEL_DEVICETREE ?= "my_machine.dtb"` inside `my_machine.conf`
- Made `SRC_URI = " \ file://my_machine.dtb \"` inside `my_machine-devicetree_1.0.bb` (which is a replica of [tegrademo-devicetree_1.0.bb](https://github.com/OE4T/tegra-demo-distro/commit/2215ed0c77139595d22c504df1e8ddb01c4b2b1d#diff-38ca49374ae162b0416744363ea7aec9485a93071202db344df3efcc89c8dac1))

Error: 

make[2]: *** No rule to make target 'arch/arm64/boot/dts/my_machine.dtb'.  Stop.
make[1]: *** [build/tmp/work-shared/my_machine/kernel-source/Makefile:1381: my_machine.dtb] Error 2
make: *** [build/tmp/work-shared/my_machine/kernel-source/Makefile:203: __sub-make] Error 2
ERROR: oe_runmake failed
ERROR: Task (poky/../meta-tegra/recipes-kernel/linux/linux-tegra_5.10.bb:do_compile) failed with exit code '1'

Seems like yocto needs to see something more to correctly incorporate my_machine.dtb.

**Reply by madisongh** [reply]
You'll need to change your recipe to omit the compilation step completely and change the installation step to directly install the pre-compiled DTB files into the right locations.

**Comment by albertbu** [comment]
HI, 
what about using a SEPARATE overlay device tree? How can this be included and used by the image at boot? Or, to clearer, how can this be applied at boot time directly?
Thanks

**Reply by madisongh** [reply]
See [this wiki page](https://github.com/OE4T/meta-tegra/wiki/Using-device-tree-overlays#runtime-application-of-overlays).

**Comment by luqtd** [comment]
Is there any way to flash only the BCT pinmux dts? Using ./doflash.sh script

**Comment by minicarpet** [comment]
I have the feeling that the wiki is not clear enough nor updated, indeed, [the wiki](https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE) says to copy dtsi files from `tegra-bootfiles_%.bbappend` BUT copying them into `recipes-bsp/tegra-binaries/tegra-flashvars` which lead (on my side using scarthgap) to a not found files from tegra-bootfiles.
I probably have to integrate generated pinmux files to my custom device tree but it is not really clear to me (I am new to this)
My flashvars is not applied at all even after creating tegra-flashvars bb append file.

**Comment by ichergui** [comment]
Hi @minicarpet 
I have done this is the past. Please let me know How I can help ?

**Reply by minicarpet** [reply]
Well I would appreciate 🙂.
I am using scarthgap branch (ie jetpack 36.4.4) with Jetson Orin module.
My first step were to create dtsi file using public file provided by NVIDIA.
My second step was to create a machine conf file which is based on  orin-nx.inc (I just copied this file and renamed it for now, I did that because later I would expect to have one conf file to handle both Jetson Orin and NX with our carrier board).
So following the wiki, they told to create a new flashvars file: so for now I copied the one from `work/*/tegra-flashvars/*/build` under `my_layer/recipes-bsp/tegra-binaries/tegra-flashvars/my_machine/flashvars` then I modified it using my dtsi file generated and created a bbappend with the following content:

`FILESEXTRAPATHS:prepend := "${THISDIR}/${BPN}:"`. Until here it was fine, but now I guess that some SRC_URI into bbappend is missing to get my flashvars right ?

The wiki says to copy dtsi under `recipes-bsp/tegra-binaries/tegra-flashvars`, but then says to create a `tegra-bootfiles_35.4.1.bbappend`, so I guess I should copy them under `my_layer/recipes-bsp/tegra-binaries/tegra-bootfiles`, isn't it ?

Finally I guess all these modifications modify ONLY tegraflash (ie the tar to flash the board) but NOT the image (boot), so do I need to create custom device tree ? And change Kernel to use it ?

**Reply by minicarpet** [reply]
After some research what I understood is:
- `recipes-bsp/tegra-binaries/tegra-flashvars/my_machine/flashvars` folder looks no more necessary (exist into kirkstone but not scarthgap)
- tegra-bootfiles is the only bbappend to create since tegra-flashvars is based on this
So now it seems that I did the first step, how do I use my new pinmux into the kernel now ?
if necessary I would be glad to upgrade the wiki.
